File: php/problem21.php
<?php
//solution: 31626
echo eulerProblem21() . "\n";

function eulerProblem21() {
    $d = range(0,10000);
    $d[0] = 0;
    $d[1] = 0;
    $d[2] = 1;
    for($i = 3; $i < count($d); ++$i) {
        $d[$i] = -1;
    }
    
    for($i = 3; $i < count($d); ++$i) {
        $d[$i] = sumOfFactors($i);
    }
    
    $sum = 0;
    for($i = 3; $i < count($d); ++$i) {
        if(isAmicable($d, $i)) {
            $sum += $i;
        }
    }
    
    return $sum;
}

function sumOfFactors($num) {
    $sum = 1;
    for($i = 2; $i <= sqrt($num); ++$i) {
        if($num % $i == 0) {
            $sum += $i;
            if($num / $i != $i) {
                $sum += $num / $i;
            }
        }
    }
    return $sum;
}

function isAmicable($arr, $i) {
    if($arr[$i] < count($arr)) {
        if($i == $arr[$arr[$i]] && $i != $arr[$i]) {
            return 1;
        }
        
        return 0;
    }
    
    return 0;
}

?>

File: python/problem21.py
#!/usr/bin/python
#solution: 31626

import math

def proper_divisors(n):
    divs = [1]
    i = 2
    while i < math.sqrt(n):
        if n % i == 0:
            divs.append(i)
            divs.append(n / i)
        i += 1
    return divs

def euler_problem_21():
    totalsum = 0
    for i in xrange(2, 10000):
        divsum = sum(proper_divisors(i))
        othersum = sum(proper_divisors(divsum))
        if i == othersum and i != divsum:
            totalsum += i
            totalsum += divsum
    return totalsum / 2 # everything gets added twice

print euler_problem_21()
